content/en/docs/upgrade/devops-data-migration-upgrade.md

```diff
@@ -0,0 +1,104 @@
+
+## DevOps Architecture Comparison
+
+### DevOps in KubeSphere v2.1
+
+![devops-2-1-architecture](/images/docs/upgrade/devops-data-migration-upgrade/devops-2-1-architecture.png)
+
+In KubeSphere v2.1, the DevOps data are stored in Jenkins, while the MySQL in KubeSphere stores the authorization information and the corresponding relations among projects, pipelines, and personnel. Frontend users make calls to the Api Gateway to get authentication from the ks-apiserver and then connect to Jenkins.
+
+Another special type is `/kapis/jenkins.kubesphere.io`, which directly passes to Jenkins through the Api Gateway and mainly used when downloading the archive files.
+
+### DevOps in KubeSphere v3.0
+
+![devops-3-0-architecture](/images/docs/upgrade/devops-data-migration-upgrade/devops-3-0-architecture.png)
+
+In KubeSphere v3.0, the DevOps data are stored in CustomResourceDefinition (CRD). The data in the CRD are synchronized to Jenkins in an one-way direction, and any previous data with the same name will be overwritten. There are mainly two types of operation. One type is the operation of creation, and the other is the operation of triggering.
+
+The operation of creation mainly involves three CRD types, including DevOps projects, pipelines and credentials. Users make calls to the interface of ks-apiserver through frontend to create resources and store them in the etcd, and then the ks-controller-manager keeps synchronizing those objects to Jenkins. The operation of triggering mainly involves instant actions, including running a pipeline and reviewing a pipeline. Users make calls to the ks-apiserver through frontend, and then, after data conversion, directly make calls to the Jenkins API.
+
+## The Data Migration Logic
+
+When upgrading KubeSphere from v2.1 to v3.0, the process of data migration follows the steps shown below:
+
+1. Retrieve information about DevOps directories from MySQL
+2. Retrieve information about pipelines and credentials from Jenkins
+3. Serialize the above information into a YAML file
+4. Back up the YAML resource object to the MinIO object storage
+5. Use the `kubectl apply` command to apply those pipeline resources
+6. Authorizations migration
+
+Data comparison before and after the migration:
+
+| Resource Affected |  v2.1   | v3.0 |
+| :---------------: | :-----: | :--: |
+|  DevOps projects  |  MySQL  | CRD  |
+|     Pipelines     | Jenkins | CRD  |
+|    Credentials    | Jenkins | CRD  |
+|  Authorizations   |  MySQL  | CRD  |
+
+{{< notice warning>}}
+
+You must back up the data in advance because the data in Jenkins will be overwritten by the ks-controller-manager once the CRD is created during data migration. You can use Velero to make a PVC-level backup, or back up the `/var/jenkins_home` directory of the Jenkins Pod directly.
```

Comment:
->
You must back up the data in advance because the data in Jenkins will be overwritten by ks-controller-manager once the CRDs are created during data migration. You can use Velero to make a PVC-level backup, or back up the `/var/jenkins_home` directory of the Jenkins Pod directly.